pageserver/src/http/routes.rs

```diff
@@ -1521,6 +1523,60 @@ async fn timeline_gc_unblocking_handler(
     block_or_unblock_gc(request, false).await
 }
 
+async fn timeline_page_trace_handler(
+    request: Request<Body>,
+    _cancel: CancellationToken,
+) -> Result<Response<Body>, ApiError> {
+    let tenant_shard_id: TenantShardId = parse_request_param(&request, "tenant_shard_id")?;
+    let timeline_id: TimelineId = parse_request_param(&request, "timeline_id")?;
+    let state = get_state(&request);
+
+    let size_limit =
+        parse_query_param::<_, u64>(&request, "size_limit_bytes")?.unwrap_or(1024 * 1024);
+    let time_limit_secs = parse_query_param::<_, u64>(&request, "time_limit_secs")?.unwrap_or(5);
+
+    // Convert size limit to event limit based on known serialized size of an event
+    let event_limit = size_limit / 32;
+
+    let timeline =
+        active_timeline_of_active_tenant(&state.tenant_manager, tenant_shard_id, timeline_id)
+            .await?;
+
+    let (page_trace, mut trace_rx) = PageTrace::new(event_limit);
+    timeline.page_trace.store(Arc::new(Some(page_trace)));
+
+    let mut buffer = Vec::with_capacity(size_limit as usize);
+
+    let deadline = Instant::now() + Duration::from_secs(time_limit_secs);
+
+    loop {
+        let timeout = deadline.saturating_duration_since(Instant::now());
+        tokio::select! {
+            event = trace_rx.recv() => {
+                buffer.extend(bincode::serialize(&event).unwrap());
+
+                if buffer.len() >= size_limit as usize {
+                    // Size threshold reached
+                    break;
+                }
+            }
+            _ = tokio::time::sleep(timeout) => {
+                // Time threshold reached
+                break;
+            }
+        }
+    }
+
+    // Above code is infallible, so we guarantee to switch the trace off when done
+    timeline.page_trace.store(Arc::new(None));
```

Comment:
nit: we could also stream to the client, and cancel if the client goes away.

Reply:
Streaming is kind of annoying with Hyper, just kept it simple. Added cancellation on disconnect though.